topi/python/topi/x86/conv2d_transpose.py

```diff
@@ -16,50 +16,76 @@
+    _, out_c, filter_h, filter_w = kernel.shape
+    stride_h, stride_w = strides
+    # dilate data
+    data_dilate = dilate(data, [1, 1, stride_h, stride_w], name='data_dilate')
+    # pad data
+    fpad_top, fpad_left, fpad_bottom, fpad_right = get_pad_tuple(padding, (filter_h, filter_w))
+    bpad_top = filter_h - 1 - fpad_top
+    bpad_bottom = filter_h - 1 - fpad_bottom
+    bpad_left = filter_w - 1 - fpad_left
+    bpad_right = filter_w - 1 - fpad_right
+    data_pad = pad(data_dilate, \
+                   [0, 0, bpad_top, bpad_left], \
+                   [0, 0, bpad_bottom, bpad_right], \
+                   name='data_pad')
+    # transform kernel layout from IOHW to OIHW, and rotate kernel by 180 degrees
+    kernel_transform = tvm.compute((out_c, in_c, filter_h, filter_w), \
+                                   lambda o, i, h, w: kernel[i][o][filter_h-1-h][filter_w-1-w], \
+                                   name='kernel_transform')
+    # reuse conv2d implementation
+    _create_tuning_space_conv2d(cfg, data_pad, kernel_transform, strides=(1,1), padding=(0,0), \
+                                dilation=(1,1), layout="NCHW")
+    if cfg.is_fallback:
+        _get_default_config_conv2d(cfg, data_pad, kernel_transform, strides=(1,1), padding=(0,0), \
+                                   out_dtype=out_dtype, layout='NCHW')
+    return _declaration_conv_impl(cfg, data_pad, kernel_transform, strides=(1,1), padding=(0,0), \
+                                  dilation=(1,1), layout="NCHW", out_dtype=out_dtype)
+
 
 @autotvm.register_topi_schedule(generic.schedule_conv2d_transpose_nchw, 'cpu', ['direct'])
-def schedule_conv2d_transpose(cfg, outs):
+def _schedule_conv2d_transpose_nchw(cfg, outs):
     """Create schedule for tensors"""
     outs = [outs] if isinstance(outs, tvm.tensor.Tensor) else outs
     s = tvm.create_schedule([x.op for x in outs])
-    scheduled_ops = []
-
-    def traverse(op):
-        """Traverse operators from computation graph"""
-        # inline all one-to-one-mapping operators except the last stage (output)
-        if tag.is_injective(op.tag):
-            if op not in s.outputs:
-                s[op].compute_inline()
-            for tensor in op.input_tensors:
-                if isinstance(tensor.op, tvm.tensor.ComputeOp) and tensor.op not in scheduled_ops:
-                    traverse(tensor.op)
-
-        if 'conv2d_transpose_nchw' in op.tag:
-            C = op.output(0)
-
-            N, OC, OH, OW = C.op.axis
-            rc, ry, rx = C.op.reduce_axis
-
-            OH, oh = s[C].split(OH, factor=2)
-            OC, oc = s[C].split(OC, factor=32)
-            IC, ic = s[C].split(rc, factor=32)
-
-            s[C].reorder(N, OC, OH, OW, oc, IC, ry, rx, ic)
-            N = s[C].fuse(N, OC)
-            s[C].vectorize(oc)
-            s[C].parallel(N)
 
-        scheduled_ops.append(op)
+    def _callback(op):
+        # reuse conv2d schedule
+        if 'conv2d_nchw' in op.tag:
```

Comment:
Add an exception check in case the op.tag is not what we want.

Reply:
I think it's fine to not do an exception check. When the tag doesn't match, the schedule part is just skipped. And we know that this case rarely happens.

Reply:
Isee. That's fair enough.